src/path.cr

```diff
@@ -351,49 +351,30 @@ struct Path
   # Path["foo.tar.gz"].extension # => ".gz"
   # ```
   def extension : String
+    return "" if @name.bytesize < 3
     bytes = @name.to_slice
+    separators = separators().map &.ord
 
-    return "" if bytes.empty?
-
-    slice_end = bytes.size
-    current = slice_end - 1
-
-    separators = self.separators.map &.ord
-
-    # ignore trailing separators
-    while separators.includes?(bytes[current]) && current > 0
-      current -= 1
-      slice_end -= 1
+    # Ignore trailing seperators
+    offset = bytes.size - 1
+    while bytes.unsafe_fetch(offset).in? separators
+      return "" if offset == 0
+      offset -= 1
     end
 
-    # if the pattern is `foo.`, it has no extension
-    return "" if bytes[current] == '.'.ord
-
-    # position the reader at the last `.` or SEPARATOR
-    # that is not the first char
-    while !separators.includes?(bytes[current]) &&
-          bytes[current] != '.'.ord &&
-          current > 0
-      current -= 1
-    end
-
-    # if we are the beginning of the string there is no extension
-    # `/foo` and `.foo` have no extension
-    return "" unless current > 0
-
-    # otherwise we are not at the beginning, and there is a previous char.
-    # if current is '/', then the pattern is prefix/foo and has no extension
-    return "" if separators.includes?(bytes[current])
+    # Get the first occurance of a seperator or a '.' past the trailing seperators (or return "")
+    dot_index = bytes.rindex(offset: offset) { |byte| byte === '.' || byte.in? separators } || return ""
```

Comment:
The `|| return ""` could be moved into a separate expression (and potentially combined with the other failure conditions as `! dot_index`)

Reply:
I'll try improving it, just didn't want to deal with .not_nil!

Reply:
I replaced it with `|| 0` now, so the return expression handled at the bottom

Reply:
You can just add another early return for that: `return "" unless dot_index`.

Reply:
Oh, I didn't know that actually works.
I originally used `return "" if dot_index == nil`
.. but appearantly that doesn't inform the compiler it can't be Nil